File: app/src/googleplay/java/org/tasks/billing/BillingClientImpl.java
package org.tasks.billing;

import static com.google.common.collect.Lists.newArrayList;
import static com.todoroo.andlib.utility.AndroidUtilities.assertMainThread;

import android.app.Activity;
import android.content.Context;
import androidx.annotation.Nullable;
import com.android.billingclient.api.BillingClient.BillingResponse;
import com.android.billingclient.api.BillingClient.FeatureType;
import com.android.billingclient.api.BillingClient.SkuType;
import com.android.billingclient.api.BillingFlowParams;
import com.android.billingclient.api.BillingFlowParams.ProrationMode;
import com.android.billingclient.api.ConsumeResponseListener;
import com.android.billingclient.api.Purchase.PurchasesResult;
import com.android.billingclient.api.PurchasesUpdatedListener;
import com.google.common.base.Joiner;
import com.google.common.collect.Iterables;
import io.reactivex.Single;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;
import java.util.List;
import org.tasks.BuildConfig;
import org.tasks.analytics.Firebase;
import org.tasks.injection.ApplicationContext;
import timber.log.Timber;

@SuppressWarnings("all")
public class BillingClientImpl implements BillingClient, PurchasesUpdatedListener {

  public static final String TYPE_SUBS = SkuType.SUBS;

  private final Inventory inventory;
  private final Firebase firebase;
  private com.android.billingclient.api.BillingClient billingClient;
  private boolean connected;
  private OnPurchasesUpdated onPurchasesUpdated;

  public BillingClientImpl(@ApplicationContext Context context, Inventory inventory, Firebase firebase) {
    this.inventory = inventory;
    this.firebase = firebase;
    billingClient =
        com.android.billingclient.api.BillingClient.newBuilder(context).setListener(this).build();
  }

  public static String BillingResponseToString(@BillingResponse int response) {
    switch (response) {
      case BillingResponse.FEATURE_NOT_SUPPORTED:
        return "FEATURE_NOT_SUPPORTED";
      case BillingResponse.SERVICE_DISCONNECTED:
        return "SERVICE_DISCONNECTED";
      case BillingResponse.OK:
        return "OK";
      case BillingResponse.USER_CANCELED:
        return "USER_CANCELED";
      case BillingResponse.SERVICE_UNAVAILABLE:
        return "SERVICE_UNAVAILABLE";
      case BillingResponse.BILLING_UNAVAILABLE:
        return "BILLING_UNAVAILABLE";
      case BillingResponse.ITEM_UNAVAILABLE:
        return "ITEM_UNAVAILABLE";
      case BillingResponse.DEVELOPER_ERROR:
        return "DEVELOPER_ERROR";
      case BillingResponse.ERROR:
        return "ERROR";
      case BillingResponse.ITEM_ALREADY_OWNED:
        return "ITEM_ALREADY_OWNED";
      case BillingResponse.ITEM_NOT_OWNED:
        return "ITEM_NOT_OWNED";
      default:
        return "Unknown";
    }
  }

  /**
   * Query purchases across various use cases and deliver the result in a formalized way through a
   * listener
   */
  @Override
  public void queryPurchases() {
    Runnable queryToExecute =
        () -> {
          Single<PurchasesResult> purchases =
              Single.fromCallable(() -> billingClient.queryPurchases(SkuType.INAPP));
          if (areSubscriptionsSupported()) {
            purchases =
                Single.zip(
                    purchases,
                    Single.fromCallable(() -> billingClient.queryPurchases(SkuType.SUBS)),
                    (iaps, subs) -> {
                      if (iaps.getResponseCode() != BillingResponse.OK) {
                        return iaps;
                      }
                      if (subs.getResponseCode() != BillingResponse.OK) {
                        return subs;
                      }
                      iaps.getPurchasesList().addAll(subs.getPurchasesList());
                      return iaps;
                    });
          }
          purchases
              .subscribeOn(Schedulers.io())
              .observeOn(AndroidSchedulers.mainThread())
              .subscribe(this::onQueryPurchasesFinished);
        };

    executeServiceRequest(queryToExecute);
  }

  /** Handle a result from querying of purchases and report an updated list to the listener */
  private void onQueryPurchasesFinished(PurchasesResult result) {
    assertMainThread();

    // Have we been disposed of in the meantime? If so, or bad result code, then quit
    if (billingClient == null || result.getResponseCode() != BillingResponse.OK) {
      Timber.w(
          "Billing client was null or result code (%s) was bad - quitting",
          result.getResponseCode());
      return;
    }

    Timber.d("Query inventory was successful.");

    // Update the UI and purchases inventory with new list of purchases
    inventory.clear();
    add(result.getPurchasesList());
  }

  @Override
  public void onPurchasesUpdated(
      @BillingResponse int resultCode, List<com.android.billingclient.api.Purchase> purchases) {
    boolean success = resultCode == BillingResponse.OK;
    if (success) {
      add(purchases);
    }
    if (onPurchasesUpdated != null) {
      onPurchasesUpdated.onPurchasesUpdated(success);
    }
    String skus =
        purchases == null
            ? "null"
            : Joiner.on(";")
                .join(
                    Iterables.transform(purchases, com.android.billingclient.api.Purchase::getSku));
    Timber.i("onPurchasesUpdated(%s, %s)", BillingResponseToString(resultCode), skus);
    firebase.reportIabResult(resultCode, skus);
  }

  private void add(List<com.android.billingclient.api.Purchase> purchases) {
    inventory.add(Iterables.transform(purchases, Purchase::new));
  }

  @Override
  public void initiatePurchaseFlow(
      Activity activity, String skuId, String billingType, @Nullable String oldSku) {
    executeServiceRequest(
        () ->
            billingClient.launchBillingFlow(
                activity,
                BillingFlowParams.newBuilder()
                    .setSku(skuId)
                    .setType(billingType)
                    .setOldSkus(oldSku == null ? null : newArrayList(oldSku))
                    .setReplaceSkusProrationMode(ProrationMode.IMMEDIATE_WITH_TIME_PRORATION)
                    .build()));
  }

  @Override
  public void addPurchaseCallback(OnPurchasesUpdated onPurchasesUpdated) {
    this.onPurchasesUpdated = onPurchasesUpdated;
  }

  private void startServiceConnection(final Runnable executeOnSuccess) {
    billingClient.startConnection(
        new com.android.billingclient.api.BillingClientStateListener() {
          @Override
          public void onBillingSetupFinished(@BillingResponse int billingResponseCode) {
            Timber.d("onBillingSetupFinished(%s)", billingResponseCode);

            if (billingResponseCode == BillingResponse.OK) {
              connected = true;
              if (executeOnSuccess != null) {
                executeOnSuccess.run();
              }
            }
          }

          @Override
          public void onBillingServiceDisconnected() {
            Timber.d("onBillingServiceDisconnected()");
            connected = false;
          }
        });
  }

  private void executeServiceRequest(Runnable runnable) {
    if (connected) {
      runnable.run();
    } else {
      // If billing service was disconnected, we try to reconnect 1 time.
      // (feel free to introduce your retry policy here).
      startServiceConnection(runnable);
    }
  }

  /**
   * Checks if subscriptions are supported for current client
   *
   * <p>Note: This method does not automatically retry for RESULT_SERVICE_DISCONNECTED. It is only
   * used in unit tests and after queryPurchases execution, which already has a retry-mechanism
   * implemented.
   */
  private boolean areSubscriptionsSupported() {
    int responseCode = billingClient.isFeatureSupported(FeatureType.SUBSCRIPTIONS);
    if (responseCode != BillingResponse.OK) {
      Timber.d("areSubscriptionsSupported() got an error response: %s", responseCode);
    }
    return responseCode == BillingResponse.OK;
  }

  @Override
  public void consume(String sku) {
    if (!BuildConfig.DEBUG) {
      throw new IllegalStateException();
    }
    if (!inventory.purchased(sku)) {
      throw new IllegalArgumentException();
    }
    final ConsumeResponseListener onConsumeListener =
        (responseCode, purchaseToken1) -> {
          Timber.d("onConsumeResponse(%s, %s)", responseCode, purchaseToken1);
          queryPurchases();
        };

    executeServiceRequest(
        () ->
            billingClient.consumeAsync(
                inventory.getPurchase(sku).getPurchaseToken(), onConsumeListener));
  }
}

File: app/src/googleplay/java/org/tasks/billing/BillingClientImpl.kt
package org.tasks.billing

import android.app.Activity
import android.content.Context
import com.android.billingclient.api.AcknowledgePurchaseParams
import com.android.billingclient.api.BillingClient.BillingResponseCode
import com.android.billingclient.api.BillingClient.ProductType
import com.android.billingclient.api.BillingClient.newBuilder
import com.android.billingclient.api.BillingClientStateListener
import com.android.billingclient.api.BillingFlowParams
import com.android.billingclient.api.BillingFlowParams.ProductDetailsParams
import com.android.billingclient.api.BillingFlowParams.SubscriptionUpdateParams
import com.android.billingclient.api.BillingResult
import com.android.billingclient.api.ConsumeParams
import com.android.billingclient.api.Purchase.PurchaseState
import com.android.billingclient.api.PurchasesUpdatedListener
import com.android.billingclient.api.QueryProductDetailsParams
import com.android.billingclient.api.QueryPurchasesParams
import com.android.billingclient.api.consumePurchase
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.NonCancellable
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withContext
import org.tasks.BuildConfig
import org.tasks.analytics.Firebase
import org.tasks.jobs.WorkManager
import timber.log.Timber
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlin.coroutines.suspendCoroutine

class BillingClientImpl(
    @ApplicationContext context: Context?,
    private val inventory: Inventory,
    private val firebase: Firebase,
    private val workManager: WorkManager
) : BillingClient, PurchasesUpdatedListener {
    private val billingClient =
        newBuilder(context!!)
            .setListener(this)
            .enablePendingPurchases()
            .build()
    private var connected = false
    private var onPurchased: (() -> Unit)? = null

    override suspend fun getSkus(skus: List<String>): List<Sku> =
        executeServiceRequest {
            val productList = skus.map {
                QueryProductDetailsParams.Product.newBuilder()
                    .setProductId(it)
                    .setProductType(ProductType.SUBS)
                    .build()
            }
            val params = QueryProductDetailsParams.newBuilder()
                .setProductList(productList)
                .build()

            val productDetailsResult = withContext(Dispatchers.IO) {
                suspendCoroutine { cont ->
                    billingClient.queryProductDetailsAsync(params) { billingResult, productDetailsList ->
                        cont.resume(billingResult to productDetailsList)
                    }
                }
            }

            productDetailsResult.first.let {
                if (!it.success) {
                    throw IllegalStateException(it.responseCodeString)
                }
            }

            productDetailsResult.second?.map { productDetails ->
                Sku(
                    productId = productDetails.productId,
                    price = productDetails.subscriptionOfferDetails?.firstOrNull()?.pricingPhases?.pricingPhaseList?.firstOrNull()?.formattedPrice
                        ?: productDetails.oneTimePurchaseOfferDetails?.formattedPrice
                        ?: ""
                )
            } ?: emptyList()
        }

    override suspend fun queryPurchases(throwError: Boolean) = try {
        executeServiceRequest {
            withContext(Dispatchers.IO + NonCancellable) {
                val subsParams = QueryPurchasesParams.newBuilder()
                    .setProductType(ProductType.SUBS)
                    .build()
                val iapsParams = QueryPurchasesParams.newBuilder()
                    .setProductType(ProductType.INAPP)
                    .build()

                val subs = suspendCoroutine { cont ->
                    billingClient.queryPurchasesAsync(subsParams) { billingResult, purchases ->
                        cont.resume(PurchasesResult(billingResult, purchases))
                    }
                }
                val iaps = suspendCoroutine { cont ->
                    billingClient.queryPurchasesAsync(iapsParams) { billingResult, purchases ->
                        cont.resume(PurchasesResult(billingResult, purchases))
                    }
                }

                if (subs.success || iaps.success) {
                    withContext(Dispatchers.Main) {
                        inventory.clear()
                        add(subs.purchases + iaps.purchases)
                    }
                } else {
                    Timber.e("SUBS: ${subs.responseCodeString} IAPs: ${iaps.responseCodeString}")
                }
            }
        }
    } catch (e: IllegalStateException) {
        if (throwError) {
            throw e
        } else {
            Timber.e(e.message)
        }
    }

    override fun onPurchasesUpdated(
        result: BillingResult,
        purchases: List<com.android.billingclient.api.Purchase>?,
    ) {
        val success = result.success
        if (success) {
            add(purchases ?: emptyList())
            onPurchased?.invoke()
            purchases
                ?.filter { !it.isAcknowledged }
                ?.forEach {
                    firebase.reportIabResult(
                        result.responseCodeString,
                        it.products.joinToString(","),
                        it.purchaseState.purchaseStateString,
                        it.orderId ?: "",
                    )
                }
        }
        workManager.updatePurchases()
    }

    private fun add(purchases: List<com.android.billingclient.api.Purchase>) {
        inventory.add(purchases.map { Purchase(it) })
    }

    override suspend fun initiatePurchaseFlow(
        activity: Activity,
        sku: String,
        skuType: String,
        oldPurchase: Purchase?,
        onPurchased: (() -> Unit)?,
    ) {
        executeServiceRequest {
            val productList = listOf(
                QueryProductDetailsParams.Product.newBuilder()
                    .setProductId(sku)
                    .setProductType(skuType)
                    .build()
            )
            val queryParams = QueryProductDetailsParams.newBuilder()
                .setProductList(productList)
                .build()

            val productDetailsResult = withContext(Dispatchers.IO) {
                suspendCoroutine { cont ->
                    billingClient.queryProductDetailsAsync(queryParams) { billingResult, productDetailsList ->
                        cont.resume(billingResult to productDetailsList)
                    }
                }
            }

            productDetailsResult.first.let {
                if (!it.success) {
                    throw IllegalStateException(it.responseCodeString)
                }
            }

            val productDetails = productDetailsResult.second?.firstOrNull()
                ?: throw IllegalStateException("Product $sku not found")

            val productDetailsParamsBuilder = ProductDetailsParams.newBuilder()
                .setProductDetails(productDetails)

            // For subscriptions (including legacy subscriptions), we need to provide an offer token
            if (skuType == ProductType.SUBS) {
                val offerToken = productDetails.subscriptionOfferDetails?.firstOrNull()?.offerToken
                    ?: throw IllegalStateException("No offer token found for subscription $sku")
                productDetailsParamsBuilder.setOfferToken(offerToken)
            }

            val productDetailsParams = productDetailsParamsBuilder.build()

            val params = BillingFlowParams.newBuilder()
                .setProductDetailsParamsList(listOf(productDetailsParams))

            oldPurchase?.let {
                params.setSubscriptionUpdateParams(
                    SubscriptionUpdateParams.newBuilder()
                        .setOldPurchaseToken(it.purchaseToken)
                        .setSubscriptionReplacementMode(BillingFlowParams.SubscriptionUpdateParams.ReplacementMode.WITH_TIME_PRORATION)
                        .build()
                )
            }

            this@BillingClientImpl.onPurchased = onPurchased
            billingClient.launchBillingFlow(activity, params.build())
        }
    }

    override suspend fun acknowledge(purchase: Purchase) {
        if (purchase.needsAcknowledgement) {
            val params = AcknowledgePurchaseParams.newBuilder()
                .setPurchaseToken(purchase.purchaseToken)
                .build()
            withContext(Dispatchers.IO) {
                suspendCoroutine { cont ->
                    billingClient.acknowledgePurchase(params) {
                        Timber.d("acknowledge: ${it.responseCodeString} $purchase")
                        cont.resume(it)
                    }
                }
            }
        }
    }

    private suspend fun connect(): BillingResult =
        suspendCancellableCoroutine { cont ->
            billingClient.startConnection(
                object : BillingClientStateListener {
                    override fun onBillingSetupFinished(result: BillingResult) {
                        if (result.success) {
                            connected = true
                            if (cont.isActive) {
                                cont.resumeWith(Result.success(result))
                            }
                        } else {
                            connected = false
                            if (cont.isActive) {
                                cont.resumeWithException(
                                    IllegalStateException(result.responseCodeString)
                                )
                            }
                        }
                    }

                    override fun onBillingServiceDisconnected() {
                        Timber.d("onBillingServiceDisconnected()")
                        connected = false
                    }
                }
            )
        }

    private suspend fun <T> executeServiceRequest(runnable: suspend () -> T): T {
        if (!connected) {
            connect()
        }
        return runnable()
    }

    override suspend fun consume(sku: String) {
        check(BuildConfig.DEBUG)
        val purchase = inventory.getPurchase(sku)
        require(purchase != null)
        executeServiceRequest {
            val result = billingClient.consumePurchase(
                ConsumeParams.newBuilder().setPurchaseToken(purchase.purchaseToken).build(),
            )
            Timber.d("consume purchase: ${result.billingResult.responseCodeString}")
            queryPurchases(throwError = false)
        }
    }

    private data class PurchasesResult(
        val billingResult: BillingResult,
        val purchasesList: List<com.android.billingclient.api.Purchase>
    ) {
        val success: Boolean
            get() = billingResult.responseCode == BillingResponseCode.OK

        val responseCodeString: String
            get() = billingResult.responseCodeString

        val purchases: List<com.android.billingclient.api.Purchase>
            get() = purchasesList
    }

    companion object {
        const val TYPE_SUBS = ProductType.SUBS
        const val STATE_PURCHASED = PurchaseState.PURCHASED

        private val BillingResult.success: Boolean
            get() = responseCode == BillingResponseCode.OK

        val BillingResult.responseCodeString: String
            get() = when (responseCode) {
                BillingResponseCode.FEATURE_NOT_SUPPORTED -> "FEATURE_NOT_SUPPORTED"
                BillingResponseCode.SERVICE_DISCONNECTED -> "SERVICE_DISCONNECTED"
                BillingResponseCode.OK -> "OK"
                BillingResponseCode.USER_CANCELED -> "USER_CANCELED"
                BillingResponseCode.SERVICE_UNAVAILABLE -> "SERVICE_UNAVAILABLE"
                BillingResponseCode.BILLING_UNAVAILABLE -> "BILLING_UNAVAILABLE"
                BillingResponseCode.ITEM_UNAVAILABLE -> "ITEM_UNAVAILABLE"
                BillingResponseCode.DEVELOPER_ERROR -> "DEVELOPER_ERROR"
                BillingResponseCode.ERROR -> "ERROR"
                BillingResponseCode.ITEM_ALREADY_OWNED -> "ITEM_ALREADY_OWNED"
                BillingResponseCode.ITEM_NOT_OWNED -> "ITEM_NOT_OWNED"
                else -> "UNKNOWN"
            }

        val Int.purchaseStateString: String
            get() = when (this) {
                PurchaseState.UNSPECIFIED_STATE -> "UNSPECIFIED_STATE"
                PurchaseState.PURCHASED -> "PURCHASED"
                PurchaseState.PENDING -> "PENDING"
                else -> this.toString()
            }
    }
}